starcheck/state_checks.py

```diff
@@ -0,0 +1,184 @@
+import numpy as np
+from cxotime import CxoTime
+
+from astropy.table import Table, vstack
+import astropy.units as u
+from functools import lru_cache
+import numpy as np
+
+import kadi.commands as kadi_commands
+import kadi.commands.states as kadi_states
+from cxotime import CxoTime
+from chandra_maneuver import duration
+from Quaternion import Quat
+
+
+@lru_cache
+def make_man_table():
+    """
+    Calculate the time for range of manuever angles using Chandra.Maneuver
+    duration.
+    """
+    durations = []
+    q0 = Quat(equatorial=(0, 0, 0))
+
+    # Use a range of angles that sample the curve pretty well by eye
+    # The duration function is a little slow and not vectorized, so
+    # this is sparse.
+    angles = [0, 5, 10, 15, 20, 25, 35, 50, 100, 150, 180]
+    for angle in angles:
+        q1 = Quat(equatorial=(angle, 0, 0))
+        durations.append(duration(q0, q1))
+
+    return Table([durations, angles], names=["duration", "angle"])
+
+
+def calc_pcad_state_nman_sums(states):
+    """
+    For each NPNT state in a monotonically increasing list of states,
+    calculate the sum of time in NMAN before it
+    and interpolate into the man_table to calculate the equivalent
+    angle for a duration.
+    """
```

Comment:
If the last entry is NMAN, why do we care?

Reply:
This is somewhat about separation of concerns and making the function do what the function name says. Essentially what you need in the overall process is to collect the contiguous NMAN states, compute the equivalent man angle and return a table with the END of the maneuver, the total time in NMAN, and the equivalent angle. This function does not care about how that is being used later.

The subsequent code will find the maneuver chunk end time which most closely matches the NPNT start.

Reply:
Maybe I should rename the function then, because my description "For each NPNT state in a monotonically increasing list of states, calculate the sum of time in NMAN before it" is saying that this function is about calculating the NMAN time before each NPNT state and if there isn't a NPNT state (the last NMAN) it doesn't apply.  But if you want my to just reorganize with your idea about what this code should be doing instead of what this code is doing now, that's fine. I think I ended up with the code I have (manually calculating instead of directly reducing states) because indeed I wasn't sure how to responsibly and reliably handle a BRIT or NSUN in continuity without overcomplicating it in the logic.  If we're just going to ignore them that's fine.